sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -339,7 +339,9 @@ case class DataSource(
         val baseRelation =
           dataSource.createRelation(sparkSession.sqlContext, caseInsensitiveOptions)
         if (baseRelation.schema != schema) {
-          throw new AnalysisException(s"$className does not allow user-specified schemas.")
```

Comment:
Actually, why should we throw an exception here? `RelationProvider` doesn't support to specify user-specified schema and users can just omit without any performance penalties as guided in this error message.